compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java

```diff
@@ -1161,6 +1167,15 @@ private BType getExpectedTypeForInferredTypedescMember(BUnionType originalType,
             return null;
         }
 
+        // Add the original union type if all the members of the original type are present in the `types` list.
+        for (BType originalMemberType : expUnionType.getOriginalMemberTypes()) {
```

Comment:
Can we add some tests where there are intersections between members of the union? Can be done separately due to time constraints.

Reply:
Added